### PR TITLE
Return `Square\TTCache\Result` when pre-loading cache entries

### PR DESCRIPTION
**Pull request by bezhermoso**
When fetching valid cache entries for use in a cached collections, you may want to know which cache entries were found vs which cache entries weren't (or no longer valid). You may want to do this in order to load the missing entries from the DB at once via `WHERE ... IN (...)`, instead of querying individually – an N+1 anti-pattern – within the `remember(...)` clause.

```php

$ids = [...];
// Map of [<ID>] => cache key: "entity-[<ID>]"
$keys = collect(array_combine($ids, $ids))
   ->map(static fn($id) => 'entity-' . $id)
   ->toArray();

$result = $tt->load($keys);

// Map of [<ID>] => missing cache key: "entity-[<ID>]"
$missingKeys = $result->missingKeys();

$models = [];
if (!empty($missingKeys)) {
   $models = Entity::query()->whereIn('id', array_keys($missingKeys))
      ->get()
      ->keyBy('id');
}

$results = [];
foreach ($keys as $id => $cacheKey) {
     $results[$id] = $tt->remember($cachKey, 0, [], function () use ($models) {
          // Pulling from pre-loaded models from previous query. No N+1!
          $entity = $models[$id];
          return $this->serialize($model);
     });
}

```

---

Also proposing the addition of `TTCache#getLastResult(): Result` to inform callers whether or not the last `remember` call was a cache-hit or a cache-miss. This is something that can be useful for debugging, telemetry, that needs to happen _outside the callback function_.

For example, if you wish to add a HTTP header to tell whether or not the middleware-level cache was a hit or a miss:

```php

public function handle(Request $request, $next)
{
     $key = $request->uri();
     $response = $this->tt->remember($key, 0, [], function () use ($request, $next) {
         $response = $next($request);
         return $response->isSuccessful() ? $response : new BypassCache($response);
     });
     $result = $this->tt->getLastResult();
     // Set the header value to either a "hit" or "miss" depending on Result:
     $response->headers->set('x-ttcache-result', empty($result->loadedKeys()) ? 'miss' : 'hit');
     return $response;
}
```

@khepin Thoughts?

### CODE REVIEW

**Comment by bezhermoso** [comment]
(Tests incoming)

File: src/TTCache.php [hunk 1]
```diff
@@ -32,6 +32,8 @@ class TTCache
 
     protected TaggedStore $cache;
 
+    protected ?Result $lastResult;
```
Comment (khepin) [inline src/TTCache.php:35]: Knew we'd have to add something for telemetry at some point, but I don't think that's the right approach here. Can we remove telemetry related updates from this PR to get the `loaded / missing` keys update finalized on its own first.

**Review by khepin — COMMENTED** [review]
Let's remove telemetry related updates from here for now. I'd like to consider multiple approaches before settling on one. Here's some of the ones I've considered:

1. global telemetry hooks

when building TTCache, you'd give an object / some functions that get called on hit / miss. I like that this is a one and done type of thing, but we might lose granularity like "categories always hit but products never do". 

2. Additional optional `remember` params. Either only on PHP8+ where you have named params, or via an `options` array.

`$ttc->remember($key, null, $cb, ['on_hit' => fn ($stats) => ..., 'on_miss' => fn ($stats) => ...])`
`$stats` could have:
- key
- duration
- number of failed tags
- ...

3. something else

I need to go back and look at how Guzzle does that, I think they have an interesting approach there.

File: src/TTCache.php (hunk 1 above)
Comment (bezhermoso) [reply src/TTCache.php:35]: Sounds good 👍

**Comment by khepin** [comment]
4. Wrapped returns

`$ttc->remember()` would always return a wrapping object that contains the stats + the value. So you wouldn't do `return $ttc->remember(...)` anymore but `return $ttc->remember(...)->value` but you could also:

```php
$result = $ttc->remember(...);
if ($result->isHit()) {
  //...
  return $result->value;
}
```

**Comment by bezhermoso** [comment]
Not a fan of the `getLastResult` idea either, just borrowing from things like `json_last_error()`-like pattern in PHP. I 100% agree that wrapped returns is the way to go, only thing stopping me was that it's quite a breaking change.

**Comment by khepin** [comment]
*hooks*:
I like telemetry hooks for the "automatic" side of it when doing actual telemetry (sending stats). But they're less useful or at least a bit clunky to use for things like adding a header to a response after getting the value.

A question that came to mind in combining global and local hooks is what we'd do with nesting. I'd say the local hooks should only apply to their direct call and any nested calls without local hooks would go back to the global ones. IE:

```php
$ttc = new TTCache(..., ['on_hit' => fn () => echo "global hit", 'on_miss' fn () => echo "global miss"]);
$ttc->remember(..., ['on_hit' => fn () => echo "local hit", 'on_miss' fn () => echo "local miss"], function () {
    // if we got here, we'd have sent `local miss` AND `global miss` (the global AND local hooks both fire) ??
    $ttc->remember(..., [], function () {
        // if we got here, even though we're wrapped in a call that has `local`, since we haven't given hooks specific
        // to this call, we'd only get `global miss`.
        // Only the global hook is fired
    });
});
```

*wrapped results*

Do we think we'd want both wrapped results AND hooks? They seem much better for updating a value (response header) than hooks. But they're much worse for getting automated telemetry.

**Comment by bezhermoso** [comment]
I think return values should target providing results about the most immediate call, and hooks to provide data about multiple, and potentially data from deeper down. So, two different mechanisms that shine in different aspects, and I don't really think implementing both would be bad.

---

A pattern to make hooks easier to handle in both global and local context is having objects, I think:

```php


class TTCacheHookSubscriber
{
    public function onHit() { ... }
    public function onMiss() { ... }
    // other events...
}

$ttc = new TTCache(..., ['subscriber' => $globalTracker = new TTCacheHookSubscriber(...)]);

$ttc->remember($key, ..., ['subscriber' => $localTracker = new TTCacheHookSubscriber(...)]);

$localTracker->isHit($key);

$globalTracker->flushMetrics();

```

**Comment by khepin** [comment]
I think I agree. We likely only want to define an `interface` for the tracker with just `onHit` / `onMiss`. Possibly receiving the original unhashed key.

But in a separate PR.

**Review by khepin — APPROVED** [review]

**Pull request merged**